LibOS/shim/src/bookkeep/shim_vma.c

```diff
@@ -37,229 +37,327 @@
+                                   void * s, void * e)
 {
-    return test_vma_contain(tmp, addr, 1) ||
-           test_vma_contain(tmp, addr + length - 1, 1) ||
-           test_vma_startin(tmp, addr, length);
+    return vma->end > s && vma->end <= e;
 }
 
-int bkeep_shim_heap (void);
-static void __set_heap_top (void * bottom, void * top);
-
-int init_vma (void)
+/*
+ * Return true if [s, e) overlaps with the area represented by vma.
+ */
+static inline bool test_vma_overlap (struct shim_vma * vma,
+                                     void * s, void * e)
 {
-    if (!(vma_mgr = create_mem_mgr(init_align_up(VMA_MGR_ALLOC)))) {
-        debug("failed allocating VMAs\n");
-        return -ENOMEM;
-    }
-
-    debug("User space range given from PAL: %p-%p\n",
-          (void *) PAL_CB(user_address.start),
-          (void *) PAL_CB(user_address.end));
-
-    heap_bottom = (void *) PAL_CB(user_address.start);
-    if (heap_bottom + DEFAULT_HEAP_MIN_SIZE > PAL_CB(executable_range.start)
-        && heap_bottom < PAL_CB(executable_range.end)
-        && heap_top > PAL_CB(executable_range.start))
-        heap_bottom = (void *) ALIGN_UP(PAL_CB(executable_range.end));
-
-    debug("setting initial heap to %p-%p\n", heap_bottom,
-          (void *) PAL_CB(user_address.end));
-
-    __set_heap_top(heap_bottom, (void *) PAL_CB(user_address.end));
-
-    bkeep_shim_heap();
-    create_lock(vma_list_lock);
-
-    return 0;
+    return test_vma_contain(vma, s, s + 1) ||
+           test_vma_contain(vma, e - 1, e) ||
+           test_vma_startin(vma, s, e);
 }
 
 /* This might not give the same vma but we might need to
    split after we find something */
-static inline void assert_vma (void)
+static inline void __assert_vma_list (void)
 {
     struct shim_vma * tmp;
     struct shim_vma * prev __attribute__((unused)) = NULL;
 
     listp_for_each_entry(tmp, &vma_list, list) {
         /* Assert we are really sorted */
-        assert(tmp->length > 0);
-        assert(!prev || prev->addr + prev->length <= tmp->addr);
+        assert(tmp->end >= tmp->start);
+        assert(!prev || prev->end <= tmp->start);
         prev = tmp;
     }
 }
 
-static struct shim_vma * __lookup_vma (const void * addr, uint64_t len);
-static struct shim_vma * __lookup_supervma (const void * addr, uint64_t length,
-                                            struct shim_vma ** prev);
-static struct shim_vma * __lookup_overlap_vma (const void * addr, uint64_t length,
-                                               struct shim_vma ** prev);
+static inline struct shim_vma *
+__lookup_vma (void * addr, struct shim_vma ** pprev)
+{
+    struct shim_vma * vma, * prev = NULL;
+
+    listp_for_each_entry(vma, &vma_list, list) {
+        if (addr < vma->start)
+            goto none;
+        if (test_vma_contain(vma, addr, addr + 1))
+            goto out;
+
+        assert(vma->end >= vma->start);
+        assert(!prev || prev->end <= vma->start);
```

Comment:
Is it okay to have overlap between two consecutive vmas ?
I don't understand the condition of  prev->end <= vma->start. may be worth a comment somewhere on vma list definition ?

Reply:
AFAIK it's not ok in the current design and such state should be impossible to reach.

This assert says that either `prev` must be `NULL` (this is the case for the first iteration) or the next entry (`vma`) should start after `prev->end` (which means that the list is sorted). There's `<=` because `end` is exclusive.

Reply:
Yes, exactly. That's why it is an assertion.